libraries/chain/kv_context_chainbase.cpp

```diff
@@ -0,0 +1,244 @@
+#include <eosio/chain/exceptions.hpp>
+#include <eosio/chain/kv_chainbase_objects.hpp>
+#include <eosio/chain/kv_context.hpp>
+
+namespace eosio { namespace chain {
+
+   struct kv_iterator_chainbase : kv_iterator {
+      using index_type = std::decay_t<decltype(std::declval<chainbase::database>().get_index<kv_index, by_kv_key>())>;
+      using tracker_type = std::decay_t<decltype(std::declval<chainbase::database>().get_mutable_index<kv_index>().track_removed())>;
+
+      chainbase::database&       db;
+      const index_type&          idx = db.get_index<kv_index, by_kv_key>();
+      tracker_type&              tracker;
+      uint32_t&                  itr_count;
+      name                       database_id;
+      name                       contract;
+      std::vector<char>          prefix;
+      std::vector<char>          next_prefix;
+      const kv_object*           current = nullptr;
+
+      kv_iterator_chainbase(chainbase::database& db, tracker_type& tracker, uint32_t& itr_count, name database_id, name contract, std::vector<char> prefix)
+         : db{ db }, tracker{ tracker }, itr_count(itr_count), database_id{ database_id }, contract{ contract }, prefix{ std::move(prefix) } {
+
+         ++itr_count;
+         next_prefix = this->prefix;
```

Comment:
If we were willing to create a special `prefix_supremum` type that contained a single `string_view` (into `prefix`) and created overloads of `operator()` within `unsigned_blob_less` to handle `prefix_supremum` as one of its arguments, then we could still use `next_prefix` in the implementation of `kv_it_prev ` but avoid this extra copy here and also avoid having to potentially iterate up to `prefix.size()` times below.

Reply:
In fact, if we are willing to redesign the intrinsics `kv_it_create` to not retain the `prefix` in native memory (and essentially create an iterator that acts as if an empty prefix was provided) and redesign the intrinsics `kv_it_next`, `kv_it_prev`, and `kv_it_lower_bound` to all take a `prefix` which defines the range of keys the iterator can cover as a restrictive context under which that operation acts, then we could avoid needing to copy a prefix at all and dramatically lower the maximum native-side memory usage a contract could impose.

Reply:
We decided to go another (simpler) way of just disallowing prefixes longer than the current maximum key limit.